docs/docs/multipart.md

```diff
@@ -0,0 +1,261 @@
+      ),
+      Part.formData[IO](name = "description", value = "A sunset")
+    )
+  )
+)
+  .map(multipartRequest =>
+    Request[IO](
+      method = Method.POST,
+      uri = uri"http://example.com/",
+      headers = multipartRequest.headers // set the headers related to this multipart request
+    ).withEntity(multipartRequest)
+  )
+```
+
+Here's a full example with a client and a server:
+```scala mdoc:silent:nest
+import org.http4s.client.Client
+import cats.effect._
+import cats.syntax.all._
+import org.http4s._
+import org.http4s.dsl.io._
+import org.http4s.headers._
+import org.http4s.multipart._
+import org.http4s.implicits._
+
+import cats.effect.unsafe.implicits.global
+
+val routes = HttpRoutes
+  .of[IO] { case request @ POST -> Root / "multipart-form" =>
+    EntityDecoder.mixedMultipartResource[IO]().use(decoder =>
+      request.decodeWith(decoder, strict = true) { multipart =>
+        val picture = multipart.parts.find(_.name.contains("picture"))
+        val pictureSize = picture.traverse(_.body.compile.count).map(_.getOrElse(0L))
+
+        val description = multipart.parts.find(_.name.contains("description"))
+        val descriptionText = description.traverse(_.bodyText.compile.string).map(_.getOrElse(""))
+
+        (pictureSize, descriptionText)
+          .flatMapN((size, description) => 
+            Ok(s"This is a $size byte file, with the description '$description'")
+          )
+      }
+    )
+  }
+
+val client = Client.fromHttpApp(routes.orNotFound)
+val request = Multiparts.forSync[IO].flatMap(multiparts =>
+    multiparts.multipart(
+      Vector(
+        Part.fileData[IO](
+          name = "picture",
+          filename = "sunset.jpg",
+          entityBody = fs2.Stream.range[IO, Int](0, 100).map(_.toByte),
+          headers = `Content-Type`(MediaType.image.jpeg)
+        ),
+        Part.formData[IO](name = "description", value = "A sunset")
+      )
+    )
+  )
+  .map(multipartRequest =>
+    Request[IO](
+      method = Method.POST,
+      uri = uri"http://example.com/multipart-form",
+      headers = multipartRequest.headers
+    )
+      .withEntity(multipartRequest)
+  )
+
+```
+```scala mdoc
+request.flatMap(client.expect[String](_))
+  .unsafeRunSync()
+```
+
+Like [UrlForm], browsers can also submit forms in a multipart request, to use this encoding the `enctype` attribute is
+used in the `form` element, like this: `<form method="post" enctype="multipart/form-data">`. 
+
+## Streaming uploads
+
+The usage of multipart is somewhat convoluted, in part because one expects a fixed-size sequence of parts when processing
+a request (notice that `parts` is a `Vector`, not a `Stream`), this means that http4s has to get to
+the end of the request so that it knows all the parts. But this isn't the only way to upload files (although it is the 
+only way to do it in pure HTML). A simpler form of upload could just stream the binary data in the request, like so:
+
+```scala mdoc:silent:nest
```

Comment:
This example does not use multipart encoding, right, it's just an example of a streaming request and how that might be used for uploading a file?
That's a useful example to have, but I don't think people would naturally go looking for it on this page.
Perhaps we could add this example to the client documentation and link to it from here?

Also, as a minor note, I feel like the `nest` modifier eventually leads to trouble.
(perhaps this is just my past bad experience)

Reply:
@valencik I've removed the code example and left just the prose, i'd be ok with making it shorter or removing it altogether if necessary. I've also removed the use of nested, which was just for letting me redefine `routes`, etc.